### DISCUSSION TITLE
Select only 2D slices that contain label from 3D volume

### DISCUSSION THREAD

**Discussion by valecorbetta** [Project-MONAI/MONAI#5948] [Q&A] [answered]
Hi,
I'm trying to use a 2D UNet to segment brain lesions (I know that a 3D would be better but I need it to be 2D). Since the lesions are present in just a few slices and the occupy a small portion of these slices, I would like to select only the slices in which the label is present. Also, I don't want the cropped sample to be centered around the slices. My input volumes have differnt number of slices. 
I have tried to do the following:

`train_transforms = Compose (
            [
                LoadImaged(keys=keys),
                AddChanneld(keys=keys),
                NormalizeIntensityd(keys=image_keys, nonzero=True), #nonzero=True -> normalize only non-zero values
                RandCropByPosNegLabeld(keys=keys, label_key="label", image_key="image",
                                        spatial_size=(128,128,128), num_samples=32, pos=4, neg=1),
                RandSpatialCropd(keys=keys, roi_size=(96,96,96), random_center=True, random_size=False),
                RandSpatialCropSamplesd(keys=keys, roi_size=(96,96,1), num_samples=96, random_center=True, random_size=False),
                ToTensord(keys=keys)
            ]
        )`

But I encounter the following error:

> ValueError: dictionary update sequence element #0 has length 4; 2 is required

So I'm guessing I'm doing something wrong. Does anyone have any suggestions on how to achieve this?

Thanks

**Comment by KumoLiu** [comment, marked as answer]
Hi @valecorbetta, I noticed that you used `num_samples` in both `RandCropByPosNegLabeld` and `RandSpatialCropSamplesd`. After your transforms, it will create a nested list that contains 32*96 samples for each image. 
IMO, you could simply remove `RandSpatialCropSamplesd` and changed the `roi_size` in `RandSpatialCropd` to be (96, 96, 1).

Hope it can help you, thanks!

**Reply by valecorbetta** [reply]
Hi, thanks a lot for your response. I have a question: `RandSpatialCropd` with a `roi_size` of (96,96,1) would only take one slice for each of the 32 samples in output of `RandCropByPosNegLabeld`, right? I would like to take a higher number of slices for each of the 32 samples. Is there a way to do that? Thanks again.

**Reply by KumoLiu** [reply]
Hi @valecorbetta, since you set the `num_samples` equal to 96, do you want all these slices? 
I'm not sure why you need to crop several times. Since all these 32*96 samples are selected from only one image. If you just want to select the 2d slices containing the label, you could use `RandCropByPosNegLabeld(keys=keys, label_key="label", image_key="image", spatial_size=(96, 96, 1), num_samples=32*96, pos=4, neg=0)`. Does this not satisfy your needs?

**Reply by naga-karthik** [reply]
Hey @KumoLiu, I also have a similar requirement of being able to select only those slices with labels. As per your suggestion, if we use `RandCropByPosNegLabeld(keys=keys, label_key="label", image_key="image", spatial_size=(96, 96, 1), num_samples=4, pos=4, neg=0)`, am I correct in understanding that this transformation selects 4 slices per image that only have the labels (at least this is what I understand by setting `neg=0`)? What happens to the rest of the slices in the image that doesn't have any labels?  

Also, this is different from selecting slices that definitely have the labels right and discarding the rest of the image, right? I mean, doesn't this transformation raise a warning for slices that don't have any labels? 

I guess I am getting confused with how exactly this `RandCropByPosNegLabeld` function works for inputs with a lot less labels (for e.g. lesions) than the typical images. Any help is greatly appreciated!

**Reply by KumoLiu** [reply]
Hi @naga-karthik,  `RandCropByPosNegLabel` crop regions with the center being a foreground or background voxel based on the Pos Neg Ratio, so `neg=0` means all 4 slices will be cropped with foreground as the center.
If you set `neg=0`, the slices selected will definitely have the foreground, it will also give a warning if no foreground can be centered.
https://github.com/Project-MONAI/MONAI/blob/a8302eca3fc31554d4bd30fae84150766ce13472/monai/transforms/utils.py#L509-L517

Hope it can help you, Thanks!

**Reply by naga-karthik** [reply]
Ah okay, this helps! Thanks a lot for clarifying!

**Reply by valecorbetta** [reply]
Hi, sorry, I was convinced I had replied to the thread, @KumoLiu your solution worked really well. Thanks again (and sorry for the late response)

**Comment by jcohenadad** [comment]
Hi, I am also trying to select only slices that have non-empty labels. When trying the code you proposed it works perfectly well (ie: only non-empty label slices are selected):

```python
train_transforms = Compose(
    [
        LoadImaged(keys=["image", "label"]),
        EnsureChannelFirstd(keys=["image", "label"]),
        ScaleIntensityRanged(keys=["image"], a_min=0, a_max=1, b_min=0.0, b_max=1.0, clip=True),
        Spacingd(keys=["image", "label"], pixdim=(0.05, 0.05, 0.05), mode=("bilinear", "nearest")),
        RandCropByPosNegLabeld(
            keys=["image", "label"],
            image_key="image",
            label_key="label",
            spatial_size=(200, 200, 1),
            pos=1,
            neg=0,
            num_samples=1,
        ),
    ]
)
```

However, when attempting the same thing using a different leading method as per [this tutorial](https://github.com/Project-MONAI/tutorials/blob/main/modules/2d_slices_from_3d_training.ipynb) with the `monai.data.PatchIterd` and `monai.data.GridPatchDataset` functions, all slices are selected (incl. those _without_ labels in them):
```python
# Volume-level transforms for both image and segmentation
train_transforms = Compose(
    [
        LoadImaged(keys=["image", "label"]),
        EnsureChannelFirstd(keys=["image", "label"]),
        ScaleIntensityd(keys="image"),
        EnsureTyped(keys=["image", "label"]),
    ]
)
# 3D dataset with preprocessing transforms
volume_ds = CacheDataset(data=train_files, transform=train_transforms)

# Volume to patch processing
patch_func = PatchIterd(
    keys=["image", "label"], patch_size=(None, None, 1), start_pos=(0, 0, 0)  # dynamic first two dimensions
)
patch_transform = Compose(
    [
        SqueezeDimd(keys=["image", "label"], dim=-1),
        RandCropByPosNegLabeld(
            keys=["image", "label"],
            label_key="label",
            spatial_size=(200, 200),
            pos=1,
            neg=0,
            num_samples=1, 
        ),
    ]
)
patch_ds = GridPatchDataset(
    data=volume_ds, patch_iter=patch_func, transform=patch_transform, with_coordinates=False
)
shuffle_ds = ShuffleBuffer(patch_ds, buffer_size=30, seed=0)
train_loader = DataLoader(patch_ds,
                          batch_size=config['train_batch_size'],
                          num_workers=config['num_workers'],
                          pin_memory=torch.cuda.is_available())
```

I cannot wrap my head around what I am doing wrong in the 2nd scenario. Also, related question: what are the benefits of going with the `monai.data.PatchIterd` and `monai.data.GridPatchDataset` approach?

**Reply by KumoLiu** [reply]
Hi @jcohenadad, thanks for your interest here.
First, `RandCropByPosNegLabeld` doesn't mean only selecting the foreground, it means it will select the patch with the foreground as the center.
Second, `GridPatchDataset` will iterate all the data from the coordinates you specified and then apply the `patch_transform `, so at this time you already have all the slices. `RandCropByPosNegLabeld` will only be applied on the patches, not images.
https://github.com/Project-MONAI/MONAI/blob/9fd6d4ced3efb344cf3b24c51d899d8fafc882d5/monai/data/grid_dataset.py#L192-L195
In your case, maybe you could try to use `PatchDataset` and define your own `patch_func`.
https://github.com/Project-MONAI/MONAI/blob/9fd6d4ced3efb344cf3b24c51d899d8fafc882d5/monai/data/grid_dataset.py#L202

Hope it can help you, thanks!

**Reply by jcohenadad** [reply]
Thank you for your quick reply @KumoLiu ! 🙏 

After spending a few hours figuring out how `PatchDataset` works, I came up with the following code, which does work, but is very limited and has a few issues (listed as "TODO"):

```python
def volume_to_patch_remove_empty_labels(dataset):
    """Assumes slice to extract patch from is on the last dimension of the 3D image.
    """
    # TODO: Question: this function should output a sequence of len=samples_per_image, but how can I retrieve
    #   samples_per_image from inside this function? For now I've hard-coded samples_per_image=1.
    # TODO: accommodate batch>1
    patch_data = []
    # TODO: Question: Is there a slice iterator in MONAI to do this more elegantly?
    for i_z in range(dataset['label'].size()[-1]):
        # TODO: Check if OK to do the operation on the Tensor or if more efficient to convert to numpy array
        image_z = dataset['image'][:, :, :, i_z]
        label_z = dataset['label'][:, :, :, i_z]
        if label_z.sum() > 0:
            patch_data.append({'image': image_z, 'label': label_z})
            return patch_data

ds = PatchDataset(data=volume_ds,
                  patch_func=volume_to_patch_remove_empty_labels,
                  samples_per_image=1,
                  transform=None)
check_loader = DataLoader(ds, batch_size=1)
```

Presently, this code only outputs _one_ slice per volume. I presume that changing the `samples_per_image` would solve it, but how to do it if the number of non-empty slices is not known in advance? (and varies across volumes).

**Reply by jcohenadad** [reply]
I tried another approach, which consists in loading each 3D volume with nibabel, extracting the non-empty 2D slices, and building the MONAI dataset iteratively. With this approach, I end up with _all_ non-empty slices, which I then need to split between train/validation:

```python
# Iterate across image/label 3D volume, fetch non-empty slice and output a single list of image/label pair
patch_data = []
for data_dict in data_dicts:
    # i=1
    nii_image = load(data_dict['image'])
    nii_label = load(data_dict['label'])
    for i_z in range(nii_label.shape[2]):
        image_z = nii_image.get_fdata()[:, :, i_z]
        label_z = nii_label.get_fdata()[:, :, i_z]
        if label_z.sum() > 0:
            patch_data.append({'image': image_z, 'label': label_z})


def patch_func(dataset):
    """Dummy function to output a sequence of dataset of length 1"""
    return [dataset]


ds = PatchDataset(data=patch_data,
                  patch_func=patch_func,
                  samples_per_image=1,
                  transform=None)
check_loader = DataLoader(ds, batch_size=1)
# --> has all non-empty slices 😊 
# TODO: split train/validation
```

**Reply by KumoLiu** [reply]
Hi @jcohenadad, maybe you could use a `Compose` to write a `patch_func`, something like:
```
sampler = Compose([
    LoadImaged(keys=["image", "label"], image_only=True),
    RandCropByPosNegLabeld(
            keys=["image", "label"],
            image_key="image",
            label_key="label",
            spatial_size=(200, 200, 1),
            pos=1,
            neg=0,
            num_samples=n_samples,
        ),
])
```
I think this can help you outputs `n_samples` slice per volume.
Additionally, maybe you should split dataset by 3D volume otherwise some 2d slices extracted from a 3d volume may be assigned to both train and validation.

Thanks!

**Reply by jcohenadad** [reply]
Thank you for your suggestion @KumoLiu. I tried but ended up with an error (see details [here](https://github.com/ivadomed/model_seg_mouse-sc_wm-gm_t1/issues/3#issuecomment-1473818911)), which I need to take time to understand.

> Additionally, maybe you should split dataset by 3D volume otherwise some 2d slices extracted from a 3d volume may be assigned to both train and validation.

While in general I agree with this comment, in the particular application I am working on, the goal is to segment a small (n=10) dataset of hundreds of slices per volume, the ground truth is only present in a few slices per volume, and volumes are fairly similar with each other, therefore it will be more efficient to mix subjects between train/eval, as opposed to separating them. The goal here is not to achieve good generalization, but simply to perform the segmentation on this specific dataset.

**Reply by KumoLiu** [reply]
Hi @jcohenadad, I think the error is because you didn't add a channel dim in your label. You could add `EnsureChannelFirst` after `LoadImage`.
Thanks!

**Reply by jcohenadad** [reply]
Thank you @KumoLiu , I am probably not using it properly: https://github.com/ivadomed/model_seg_mouse-sc_wm-gm_t1/issues/3#issuecomment-1474069925

**Reply by KumoLiu** [reply]
Could you show the whole error message? Thanks!

**Reply by KumoLiu** [reply]
Just find your source code, I think you should use `EnsureChannelFirstd(keys=["image", "label"])` instead of `EnsureChannelFirst()`. Thanks!

**Reply by jcohenadad** [reply]
Good catch! The issue is indeed now [solved](https://github.com/ivadomed/model_seg_mouse-sc_wm-gm_t1/pull/9/commits/4042dc744c74c08b0e05976b513b1bb8c1b1de50). Thank you! While continuing to investigate the parallel approach "A" described in https://github.com/Project-MONAI/MONAI/discussions/5948#discussioncomment-5326485 (latest working prototype at https://github.com/ivadomed/model_seg_mouse-sc_wm-gm_t1/blob/main/train.py), I can try the approach "B" proposed by @KumoLiu in https://github.com/ivadomed/model_seg_mouse-sc_wm-gm_t1/blob/jca/loader-test/train.py. But given that in this particular application, I would like to mix train/validation volumes, is there an actual incentive to work with "B" vs. "A"?

**Reply by KumoLiu** [reply]
Didn't get your questions, did you mean `PatchDataset` and `GridPatchDataset`?

**Reply by jcohenadad** [reply]
Sorry for not being clear. Here are my questions:
- With [this approach](https://github.com/ivadomed/model_seg_mouse-sc_wm-gm_t1/blob/4042dc744c74c08b0e05976b513b1bb8c1b1de50/train.py#L158-L178), the split is done on the 3D volumes, which in this case is undesired. Instead, I would like to extract all non-empty 2D slices across all volumes, and _then_ do the train/validation split. Is there a way to split a "DataLoader()" object?
- There is no fixed number of non-empty 2D slices per 3D volume (eg: volume 1 has 11, volume 2 has 8, etc.), therefore, setting `samples_per_image` to a fixed number would not work. Is there a _dynamic_ way to set `samples_per_image` for each volume?
Thanks!
